nginx/src/charmlibs/nginx/config.py

```diff
@@ -0,0 +1,622 @@
+            ]
+
+        @property
+        def _upstream_addresses(self) -> Dict[str, Set[str]]:
+            # a mapping from an upstream "role" to the set of addresses
+            # that belong to this upstream
+            return {
+                "upstream1": {"address1", "address2"},
+                "upstream2": {"address3", "address4"},
+            }
+
+        @property
+        def _tls_available(self) -> bool:
+            # return if the Nginx config should have TLS enabled
+            pass
+
+        def _reconcile(self):
+            if self._container.can_connect():
+                new_config: str = self._nginx.get_config(self._upstream_addresses,
+                  self._tls_available)
+                should_restart: bool = self._has_config_changed(new_config)
+                self._container.push(self.config_path, new_config, make_dirs=True)
+                self._container.add_layer("nginx", self.layer, combine=True)
+                self._container.autostart()
+
+                if should_restart:
+                    logger.info("new nginx config: restarting the service")
+                    self.reload()
+
+        def _nginx_upstreams(self) -> List[NginxUpstream]:
+            # UPSTREAMS is a list of backend services that we want to route traffic to
+            for upstream in UPSTREAMS:
+                # UPSTREAMS_PORT is the port the backend services are running on
+                upstreams.append(NginxUpstream(upstream, UPSTREAMS_PORT, upstream))
+                return upstreams
+
+        def _server_ports_to_locations(self) -> Dict[int, List[NginxLocationConfig]]:
+            # NGINX_PORT is the port an nginx server is running on
+            # Note that: you can define multiple server blocks, each running on a different port
+            return {NGINX_PORT: self._nginx_locations}
+
+```
+"""
+
+import logging
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Any, Final, Literal, cast
+
+import crossplane
+
+from charmlibs.nginx.tls_config_mgr import CERT_PATH, KEY_PATH
+
+from .utils import is_ipv6_enabled
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TLS_VERSIONS: Final[list[str]] = ['TLSv1', 'TLSv1.1', 'TLSv1.2', 'TLSv1.3']
+
+
+# Define valid Nginx `location` block modifiers.
+# cfr. https://www.digitalocean.com/community/tutorials/nginx-location-directive#nginx-location-directive-syntax
+NginxLocationModifier = Literal[
+    '',  # prefix match
+    '=',  # exact match
+    '~',  # case-sensitive regex match
+    '~*',  # case-insensitive regex match
+    '^~',  # prefix match that disables further regex matching
+]
+
+
+@dataclass
+class NginxLocationConfig:
+    """Represents a `location` block in an Nginx configuration.
+
+    For example,
+    ``NginxLocationConfig('/', 'foo', backend_url="/api/v1" headers={'a': 'b'},
+      modifier=EXACT, is_grpc=True, use_tls=True)``
+
+    would result in the nginx config:
```

Comment:
This should fix the docs build. The `::` and following empty lines before indented code sections are the important bit.

```suggestion
    For example::

        NginxLocationConfig(
            '/',
            'foo',
            backend_url="/api/v1"
            headers={'a': 'b'},
            modifier=EXACT,
            is_grpc=True,
            use_tls=True,
        )

    would result in the nginx config::

```